### PR TITLE
add docker for mongo

### PR DESCRIPTION
**Pull request by ayz-lex**
Restructured repository such that there are two folders that contain the dockerfiles; one is for mongo, and one is for postgres. Changed README to include information about setting up the mongo docker, as well as more clarity.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -17,64 +54,60 @@ manage.py requires Docker to be installed. Search online to learn how to install
 - `sudo systemctl start docker.service`
 - `sudo systemctl enable docker.service`
 - You might need to restart the docker service/relogin your user account.
-## Summary
-This Dockerfile specifies a container that builds upon the default "postgres" container in Docker Hub.
+# Summary
+
+## Top
```
Comment (vedantroy) [inline README.md]: Nitpick: rename top to intro? Idk something besides "top".

**Comment by vedantroy** [comment]
@ayz-lex . Try to reduce code duplication and duplication in the README. 

The argument parser automatically rejects invalid choices (i.e -- they can't specify `mysql` instead of `postgres` or `mongo`). This means you don't need if statements to handle whether the option is `mongo` or `postgres` when building the container, you can just prepend `buildUMass/` to the input.

Also, change the code for tagging images: https://docs.docker.com/engine/reference/commandline/build/#tag-an-image--t

Run `docker images`, you'll see that right now the "repository" is "mongo-image". That doesn't seem correct. The repository should be "build-umass" (or something similar). The image name should just be "mongo", or maybe "easy-mongo". "mongo-image" is redundant because we already know it's an image.

File: README.md (hunk 1 above)
Comment (ayz-lex) [reply README.md]: lol in debate we used top as in "top level"

Comment (ayz-lex) [reply README.md]: will change though

**Comment by ayz-lex** [comment]
> @ayz-lex . Try to reduce code duplication and duplication in the README.
> 
> The argument parser automatically rejects invalid choices (i.e -- they can't specify `mysql` instead of `postgres` or `mongo`). This means you don't need if statements to handle whether the option is `mongo` or `postgres` when building the container, you can just prepend `buildUMass/` to the input.
> 
> Also, change the code for tagging images: https://docs.docker.com/engine/reference/commandline/build/#tag-an-image--t
> 
> Run `docker images`, you'll see that right now the "repository" is "mongo-image". That doesn't seem correct. The repository should be "build-umass" (or something similar). The image name should just be "mongo", or maybe "easy-mongo". "mongo-image" is redundant because we already know it's an image.

Not sure how to comment exactly, so might be doing this wrong. I understand the last two paragraphs. I don't understand the first paragraph, of the point that checking invalid choices is unnecessary. Are you saying "elif" is not necessary, and should be changed to "else" instead? Also can be elaborate on the buildUMass/ part (last sentence).

**Comment by vedantroy** [comment]
@ayz-lex Have something like this
```
if args.build:
  # at this point, args.build_choice is either "postgres" or "mongo". Utilize that fact.
  build_image(args.build_choice)
```

build_image:
```
def build_image(location):
  image_tag = f"buildUMass/easy-{location}"
  rest of the code...
```

File: manage.py [hunk 2]
```diff
@@ -33,31 +34,31 @@ def easy_exec(cmd):
...
   if args.volume:
     print("Warning: irrevlant flag -v/--volume")
   if args.image:
     print("warning: irrelevant flag -i/--image")
-  cmd = f"docker build -t image-name ./" + loc
+  cmd = f"docker build -t {image_tag} ./{location}"
   easy_exec(cmd)
 
-def start(nm, vol, img, dbroute, port):
+def start(name, volume, image, volume_location, port_binding):
```
Comment (vedantroy) [inline manage.py:48]: Rename `volume_location` to `volume_location_in_container` (longer but clearer).

Rename `name` to `container_name`

Comment (vedantroy) [inline manage.py]: Rename `volume_location` to `volume_location_in_container`. Rename `name` to `container_name`

File: manage.py [hunk 3]
```diff
@@ -68,14 +69,14 @@ def stop(nm):
...
   elif args.build_choice == 'postgres':
-    build(f"postgres")    
+    build_image("postgres-image", "postgres")    
 elif args.sub == 'start':
   if args.start_choice == 'mongo':
-    start("mongo_docker", "mongo-volume", f"image-name:latest", f"/data/db", f"27017:27017")
+    start("mongo_docker", "mongo-volume", "mongo-image:latest", "/data/db", "27017:27017")
   elif args.start_choice == 'postgres':
-    start("pg_docker", "PGDATA", f"image-name:latest", f"/var/lib/postgresql/data", f"5432:5432")
+    start("pg_docker", "PGDATA", "postgres-image:latest", "/var/lib/postgresql/data", "5432:5432")
```
Comment (vedantroy) [inline manage.py]: If we use `mongo-volume`, then the postgres volume should be called `postgres_volume`

File: manage.py [hunk 4]
```diff
@@ -68,14 +69,14 @@ def stop(nm):
...
 if args.sub == 'build':
   if args.build_choice == 'mongo':
-    build(f"mongo")
+    build_image("mongo-image", "mongo")
   elif args.build_choice == 'postgres':
-    build(f"postgres")    
+    build_image("postgres-image", "postgres")    
 elif args.sub == 'start':
   if args.start_choice == 'mongo':
-    start("mongo_docker", "mongo-volume", f"image-name:latest", f"/data/db", f"27017:27017")
```
Comment (vedantroy) [inline manage.py]: Give an optional flag, so the user can choose what port on the host machine is bound to the port inside the container. As usual, there should be sensible defaults (`27107` for mongo, `5432` for postgres).

File: README.md [hunk 5]
```diff
@@ -22,66 +54,60 @@ manage.py requires Docker to be installed. Search online to learn how to install
...
-- `-p 5432:5432` Bind port 5432 on the host to port 5432 on the container, which is the default port that Postgres listens on. The syntax is HOST_PORT:CONTAINER_PORT
-- `-- ip localhost` Set the container's ip address to localhost. By default, it is 0.0.0.0. Thus, programs can connect to the Postgres database at `localhost:5432` instead of `0.0.0.0:5432`.
-- `<image id>` The id of the Docker image that this Container will be constructed/instantiated from.
+**For MongoDB:**
 
-**Connect to the Postgres Database**: `psql -h localhost -p 5432 -U dev_user -d dev_db`
-- `-p 5432` This flag is optional since by default psql will use port 5432.
+In the container, Mongo uses a folder "/db/data" to store all the data that is created in this container for later use (i.e. after we stop the container, and start it again), we must have to create a location on our local machine that is able to retain such data. Hence, we have to bind a location on our local machine, to that directory in the container. Thankfully, Docker has an option to do this for us, by managing the database files on our local machine. Instead of binding a specific directory, Docker will mount a volume for us.
```
Comment (vedantroy) [inline README.md]: Merge the mongo and postgres section in this into 1. There's a lot of copy and paste in the readme.

File: manage.py [hunk 6]
```diff
@@ -2,55 +2,83 @@
...
+  print("------")
+  retval = p.wait()
+  if retval != 0:
+    print(f"Error executing command. The process returned {retval}.")
+  return retval == 0
 
-assert xor(args.start, args.kill)
+def build_image(image_tag, location):
+  print(f"Building image with tag: {image_tag}")
+  if args.name:
```
Comment (vedantroy) [inline manage.py:42]: Make a function called `check_irrelevant_argument`

Something like
```python3
def irrelevant_argument(arg):
  # convert args to dictionary
  args_dict = vars(args)
  if args_dict[arg]:
   # arg[0] is the first character in the string
    print(f"Warning: irrelevant flag -{arg[0]}/--{arg}")
```

Of course, test this code before you use it.

Comment (ayz-lex) [reply manage.py:42]: i don't think this is necessary

**Pull request closed without merging**